File: 3507. minimum-pair-removal-to-sort-array-i.java
import java.util.Arrays;
import java.util.List;
import java.util.ArrayList;

class Solution {
    public int minimumPairRemoval(List<Integer> nums) {
        // Create a mutable copy as ArrayList
        ArrayList<Integer> arr = new ArrayList<>(nums);
        
        int ans = 0;
        int n = arr.size();
        
        while (!isSorted(arr, n)) {
            ans++;
            int minSum = Integer.MAX_VALUE;
            int pos = -1;
            
            for (int i = 1; i < n; i++) {
                int sum = arr.get(i - 1) + arr.get(i);
                if (sum < minSum) {
                    minSum = sum;
                    pos = i;
                }
            }
            
            arr.set(pos - 1, minSum);
            arr.remove(pos);
            n--;
        }
        
        return ans;
    }
    
    private boolean isSorted(ArrayList<Integer> nums, int n) {
        for (int i = 1; i < n; i++) {
            if (nums.get(i) < nums.get(i - 1)) {
                return false;
            }
        }
        return true;
    }

    // Test cases
    public static void main(String[] args) {
        Solution solution = new Solution();
        
        // Test case 1
        List<Integer> nums1 = Arrays.asList(5, 2, 3, 1);
        int output1 = solution.minimumPairRemoval(nums1);
        System.out.println("Test 1: nums = [5,2,3,1]");
        System.out.println("Output: " + output1 + ", Expected: 2");
        
        // Test case 2
        List<Integer> nums2 = Arrays.asList(1, 2, 2);
        int output2 = solution.minimumPairRemoval(nums2);
        System.out.println("Test 2: nums = [1,2,2]");
        System.out.println("Output: " + output2 + ", Expected: 0");
        
        System.out.println("\nTest cases executed!");
    }
}

File: 3507. minimum-pair-removal-to-sort-array-i.py
from typing import List


class Solution:
    def minimumPairRemoval(self, nums: List[int]) -> int:
        def isSorted(nums, n) -> bool:
            for i in range(1,n):
                if nums[i] < nums[i - 1]: return False
            return True
        ans, n = 0, len(nums)
        while not isSorted(nums, n):
            ans += 1
            min_sum, pos = float('inf'), -1
            for i in range(1,n):
                sum = nums[i - 1] + nums[i]
                if sum < min_sum:
                    min_sum = sum
                    pos = i
            nums[pos - 1] = min_sum
            for i in range(pos, n-1): nums[i] = nums[i + 1]
            n -= 1
        return ans


# Test cases
if __name__ == "__main__":
    solution = Solution()
    
    # Test case 1
    nums1 = [5,2,3,1]
    output1 = solution.minimumPairRemoval(nums1)
    print(f"Test 1: nums = [5,2,3,1]")
    print(f"Output: {output1}, Expected: 2")
    
    # Test case 2
    nums2 = [1,2,2]
    output2 = solution.minimumPairRemoval(nums2)
    print(f"Test 2: nums = [1,2,2]")
    print(f"Output: {output2}, Expected: 0")
    
    print("\nTest cases executed!")
 